src/ray/common/id.h

```diff
@@ -78,78 +96,209 @@ class UniqueID : public BaseID<UniqueID> {
+  ///
+  /// \return The maximum index of object.
+  static uint64_t MaxObjectIndex() { return kMaxObjectIndex; }
+
+  static size_t Size() { return kLength; }
+
+  /// Generate ObjectID by the given binary string of a plasma id.
+  ///
+  /// \param from The binary string of the given plasma id.
+  /// \return The ObjectID converted from a binary string of the plasma id.
+  static ObjectID FromPlasmaIdBinary(const std::string &from);
+
+  /// Generate an random actor cursor object id by the given actor.
+  ///
+  /// \param actor_id TODO(qwang)
+  /// \return
+  static ObjectID GenerateActorDummyObjectId(const ActorID &actor_id);
+
   plasma::ObjectID ToPlasmaId() const;
+
   ObjectID(const plasma::UniqueID &from);
 
   /// Get the index of this object in the task that created it.
   ///
   /// \return The index of object creation according to the task that created
-  /// this object. This is positive if the task returned the object and negative
-  /// if created by a put.
-  int32_t ObjectIndex() const { return index_; }
+  /// this object.
+  ObjectIDIndexType ObjectIndex() const;
 
   /// Compute the task ID of the task that created the object.
   ///
   /// \return The task ID of the task that created this object.
   TaskID TaskId() const;
 
+  /// Whether this object is created by a task.
+  ///
+  /// \return True if this object is created by a task, otherwise false.
+  bool CreatedByTask() const;
+
+  /// Whether this object was created through `ray.put`.
+  ///
+  /// \return True if this object was created through `ray.put`.
+  bool IsPutObject() const;
+
+  /// Whether this object was created as a return object of a task.
+  ///
+  /// \return True if this object is a return value of a task.
+  bool IsReturnObject() const;
+
+  /// Get the transport type of this object.
+  ///
+  /// \return The type of the transport which is used to transfer this object.
+  uint8_t GetTransportType() const;
+
   /// Compute the object ID of an object put by the task.
   ///
   /// \param task_id The task ID of the task that created the object.
   /// \param index What index of the object put in the task.
+  ///
   /// \return The computed object ID.
-  static ObjectID ForPut(const TaskID &task_id, int64_t put_index);
+  static ObjectID ForPut(const TaskID &task_id, ObjectIDIndexType put_index);
 
   /// Compute the object ID of an object returned by the task.
   ///
   /// \param task_id The task ID of the task that created the object.
   /// \param return_index What index of the object returned by in the task.
+  /// \param transport_type Which type of the transport that is used to
+  ///        transfer this object.
+  ///
   /// \return The computed object ID.
-  static ObjectID ForTaskReturn(const TaskID &task_id, int64_t return_index);
+  static ObjectID ForTaskReturn(const TaskID &task_id, ObjectIDIndexType return_index,
+                                uint8_t transport_type = 0);
```

Comment:
```suggestion
                                uint8_t transport_type);
```
There should be no notion of a default transport type.